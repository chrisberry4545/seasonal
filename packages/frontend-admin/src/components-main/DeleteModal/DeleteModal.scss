.c-delete-modal {
  position: relative;
}

.c-delete-modal__btns {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 1rem;
}

.c-delete-modal__spinner {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}
