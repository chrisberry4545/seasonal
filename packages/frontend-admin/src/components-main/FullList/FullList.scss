.c-full-list {
  position: relative;
}

.c-full-list__item {
  margin-bottom: 1rem;
}

.c-full-list__item-name {
  display: block;
}

.c-full-list-item__options {
  display: flex;
}

.c-full-list-item__edit {
  padding-left: 0;
}
