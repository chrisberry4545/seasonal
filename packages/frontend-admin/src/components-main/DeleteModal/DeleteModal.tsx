import React, { useState } from 'react';
import './DeleteModal.scss';
import { LoadingSpinner, BareButton, TextSmall, TextMedium } from '@chrisb-dev/seasonal-shared-frontend-components';
import { ModalLayout } from '../../components-layouts';

export interface IDeleteModal<T> {
  deleteItem: () => Promise<T>;
  cancel: () => void;
}

export const DeleteModal = <T extends {}> ({
  deleteItem,
  cancel
}: IDeleteModal<T>) => {
  const [isLoading, setIsLoading] = useState(false);

  const deleteButtonPressed = async () => {
    setIsLoading(true);
    await deleteItem();
    setIsLoading(false);
  };

  return (
    <ModalLayout>
      <div className='c-delete-modal'>
        <TextMedium>Are you sure you want to delete this item?</TextMedium>
        <div className='c-delete-modal__btns'>
          <BareButton onClick={cancel}>
            <TextSmall>Cancel</TextSmall>
          </BareButton>
          <BareButton onClick={(deleteButtonPressed)}>
            <TextSmall>Delete</TextSmall>
          </BareButton>
        </div>
        {
          isLoading
          && <div className='c-delete-modal__spinner'><LoadingSpinner /></div>
        }
      </div>
    </ModalLayout>
  );
};
