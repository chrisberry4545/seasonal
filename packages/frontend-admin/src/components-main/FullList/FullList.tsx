import React, { useState } from 'react';
import { FormLayout } from '../../components-layouts';
import { TextHeadingMedium, TextSmall, TextMedium, BareButton } from '@chrisb-dev/seasonal-shared-frontend-components';
import { Redirect } from 'react-router-dom';
import { DeleteItemButton } from '../DeleteItemButton/DeleteItemButton';
import './FullList.scss';

export interface IFullListProps<T> {
  title: string;
  items: T[] | undefined;
  getItemName: (item: T) => string;
  getItemEditUrl: (item: T) => string;
  deleteItemFunc: (item: T) => Promise<T>;
}

export const FullList = <T extends {}>({
  title,
  items,
  getItemName,
  getItemEditUrl,
  deleteItemFunc
}: IFullListProps<T>) => {
  const [goToEditLink, setGoToEditLink] = useState<null | string>(null);

  if (goToEditLink) {
    return <Redirect to={goToEditLink} push={true} />;
  }

  return <FormLayout>
    <TextHeadingMedium className='c-full-list__heading'>{title}</TextHeadingMedium>
    {
      items && items.map((item) =>
        <div className='c-full-list__item' key={getItemEditUrl(item)}>
          <TextMedium className='c-full-list__item-name'>{getItemName(item)}</TextMedium>
          <div className='c-full-list-item__options'>
            <BareButton className='c-full-list-item__edit'
              onClick={() => setGoToEditLink(getItemEditUrl(item))}>
              <TextSmall>
                Edit
              </TextSmall>
            </BareButton>
            <DeleteItemButton deleteItem={() => deleteItemFunc(item)} />
          </div>
        </div>
      )
    }
  </FormLayout>;
};
