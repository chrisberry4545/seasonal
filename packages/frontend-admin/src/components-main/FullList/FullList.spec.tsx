
import React from 'react';
import { FullList } from './FullList';
import { shallow, ShallowWrapper } from 'enzyme';
import { BareButton } from '@chrisb-dev/seasonal-shared-frontend-components';
import { DeleteModal } from '../DeleteModal/DeleteModal';

jest.mock('@chrisb-dev/seasonal-shared-frontend-components', () => ({
  BareButton: () => 'BareButton',
  TextMedium: () => 'TextMedium',
  TextSmall: () => 'TextSmall'
}));
jest.mock('../DeleteModal/DeleteModal', () => ({
  DeleteModal: () => 'DeleteModal'
}));
jest.mock('../../components-layouts', () => ({
  LayoutWithTitle: () => 'LayoutWithTitle'
}));

describe('<FullList />', () => {
  let wrapper: ShallowWrapper;
  let mockGetItemId: jest.Mock;
  let mockGetItemName: jest.Mock;
  let mockGetItemEditUrl: jest.Mock;
  let mockDelete: jest.Mock;
  const items = [{}, {}];

  beforeEach(() => {
    mockGetItemId = jest.fn().mockReturnValue('id');
    mockGetItemName = jest.fn().mockReturnValue('name');
    mockGetItemEditUrl = jest.fn().mockReturnValue('edit-url');
    mockDelete = jest.fn();
    wrapper = shallow(
      <FullList
        title='Title'
        items={items}
        getItemId={mockGetItemId}
        getItemName={mockGetItemName}
        getItemEditUrl={mockGetItemEditUrl}
        deleteItemFunc={mockDelete} />
    );
  });

  test('renders correctly', () => expect(wrapper).toMatchSnapshot());

  test('can click edit', () => {
    const onClick = wrapper.find(BareButton).first().props().onClick;
    if (onClick) {
      onClick();
    }
    expect(mockGetItemEditUrl).toHaveBeenCalled();
  });

  describe('when editing an item', () => {
    beforeEach(() => {
      const onClick = wrapper.find(BareButton).first().props().onClick;
      if (onClick) {
        onClick();
      }
    });

    test('redirects to the edit url', () => expect(wrapper).toMatchSnapshot());

  });

  describe('when deleting the item', () => {
    beforeEach(() => {
      const onClick = wrapper.find(BareButton).last().props().onClick;
      if (onClick) {
        onClick();
      }
    });

    test('shows a delete confirmation modal', () =>
      expect(wrapper.find(DeleteModal).exists()).toBe(true));

    test('can close the modal', () => {
      wrapper.find(DeleteModal).first().props().cancel();
      expect(wrapper.find(DeleteModal).exists()).toBe(false);
    });

    test('can delete the item', () => {
      wrapper.find(DeleteModal).first().props().deleteItem();
      expect(mockDelete).toHaveBeenCalledWith(items[0]);
    });

  });

});
