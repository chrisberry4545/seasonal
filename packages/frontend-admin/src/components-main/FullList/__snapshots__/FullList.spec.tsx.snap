// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FullList /> renders correctly 1`] = `
<div
  className="c-full-list"
>
  <LayoutWithTitle
    title="Title"
  >
    <div
      className="c-full-list__item"
      key="id"
    >
      <TextMedium
        className="c-full-list__item-name"
      >
        name
      </TextMedium>
      <div
        className="c-full-list-item__options"
      >
        <BareButton
          className="c-full-list-item__edit"
          onClick={[Function]}
        >
          <TextSmall>
            Edit
          </TextSmall>
        </BareButton>
        <BareButton
          onClick={[Function]}
        >
          <TextSmall>
            Delete
          </TextSmall>
        </BareButton>
      </div>
    </div>
    <div
      className="c-full-list__item"
      key="id"
    >
      <TextMedium
        className="c-full-list__item-name"
      >
        name
      </TextMedium>
      <div
        className="c-full-list-item__options"
      >
        <BareButton
          className="c-full-list-item__edit"
          onClick={[Function]}
        >
          <TextSmall>
            Edit
          </TextSmall>
        </BareButton>
        <BareButton
          onClick={[Function]}
        >
          <TextSmall>
            Delete
          </TextSmall>
        </BareButton>
      </div>
    </div>
  </LayoutWithTitle>
</div>
`;

exports[`<FullList /> when editing an item redirects to the edit url 1`] = `
<Redirect
  push={true}
  to="edit-url"
/>
`;
